src/envoy/http/authnv2/filter.cc

```diff
@@ -0,0 +1,236 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <string>
+
+#include "common/http/utility.h"
+#include "extensions/filters/http/well_known_names.h"
+#include "include/istio/utils/attribute_names.h"
+#include "src/envoy/http/authnv2/filter.h"
+#include "src/envoy/utils/authn.h"
+#include "src/envoy/utils/filter_names.h"
+#include "src/envoy/utils/utils.h"
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_split.h"
+#include "common/json/json_loader.h"
+#include "google/protobuf/struct.pb.h"
+#include "include/istio/utils/attribute_names.h"
+#include "src/envoy/http/jwt_auth/jwt.h"
+
+namespace Envoy {
+namespace Http {
+namespace Istio {
+namespace AuthN {
+
+AuthnV2Filter::~AuthnV2Filter() {}
+
+void AuthnV2Filter::onDestroy() {
+  ENVOY_LOG(debug, "Called AuthnV2Filter : {}", __func__);
+}
+
+// The JWT audience key name
+static const std::string kJwtAudienceKey = "aud";
+// The JWT issuer key name
+static const std::string kJwtIssuerKey = "iss";
+// The key name for the original claims in an exchanged token
+static const std::string kExchangedTokenOriginalPayload = "original_claims";
+
+// Extract JWT claim as a string list.
+// This function only extracts string and string list claims.
+// A string claim is extracted as a string list of 1 item.
+// A string claim with whitespace is extracted as a string list with each
+// sub-string delimited with the whitespace.
+void ExtractStringList(const std::string& key, const Envoy::Json::Object& obj,
+                       std::vector<std::string>* list) {
+  // First, try as string
+  try {
+    // Try as string, will throw execption if object type is not string.
+    const std::vector<std::string> keys =
+        absl::StrSplit(obj.getString(key), ' ', absl::SkipEmpty());
+    for (auto key : keys) {
+      list->push_back(key);
+    }
+  } catch (Json::Exception& e) {
+    // Not convertable to string
+  }
+  // Next, try as string array
+  try {
+    std::vector<std::string> vector = obj.getStringArray(key);
+    for (const std::string v : vector) {
+      list->push_back(v);
+    }
+  } catch (Json::Exception& e) {
+    // Not convertable to string array
+  }
+}
+
+// Helper function to set a key/value pair into Struct.
+static void setKeyValue(::google::protobuf::Struct& data,
+                        const std::string& key, const std::string& value) {
+  (*data.mutable_fields())[key].set_string_value(value);
+}
+
+// Returns true if the attribute populated to authn filter succeeds.
+bool ProcessJwt(const std::string& jwt, ProtobufWkt::Struct& authn_data) {
+  Envoy::Json::ObjectSharedPtr json_obj;
+  try {
+    json_obj = Json::Factory::loadFromString(jwt);
+    // ENVOY_LOG(debug, "{}: json object is {}", __FUNCTION__,
+    //           json_obj->asJsonString());
+  } catch (...) {
+    return false;
+  }
+
+  ProtobufWkt::Struct claim_structs;
```

Comment:
We currently have a hard-coded logic to handle [RcToken-type JWT ](https://github.com/istio/proxy/blob/master/src/envoy/http/authn/authenticator_base.cc#L129). Basically, if the token come from `ingress-authorization` header, we extract claims differently. I don't like that logic, and want to see how to do it better in here (we don't have the location information for this filter neither, at least at the moment).

May be leave a TODO / issue and support it in the next PR. Alternatively, check if the payload contain [`original_claims`](https://github.com/istio/proxy/blob/master/src/envoy/http/authn/authn_utils.cc#L36) to switch to the RcToken mode.